File: app/src/test/java/io/github/hidroh/materialistic/data/SessionManagerTest.java
package io.github.hidroh.materialistic.data;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.ObjectGraph;
import dagger.Provides;
import io.github.hidroh.materialistic.DataModule;
import io.github.hidroh.materialistic.test.InMemoryCache;
import rx.Scheduler;
import rx.schedulers.Schedulers;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

@RunWith(JUnit4.class)
public class SessionManagerTest {
    @Inject LocalCache cache;
    private SessionManager manager;

    @Before
    public void setUp() {
        ObjectGraph objectGraph = ObjectGraph.create(new TestModule());
        objectGraph.inject(this);
        cache.setViewed("1");
        cache.setViewed("2");
        manager = new SessionManager();
        objectGraph.inject(manager);
    }

    @Test
    public void testIsViewedNull() {
        assertFalse(manager.isViewed(null)
                .toBlocking().single());
    }

    @Test
    public void testIsViewedTrue() {
        assertTrue(manager.isViewed("1")
                .toBlocking().single());
    }

    @Test
    public void testIsViewedFalse() {
        assertFalse(manager.isViewed("-1")
                .toBlocking().single());
    }

    @Module(
            injects = {
                    SessionManagerTest.class,
                    SessionManager.class
            },
            overrides = true
    )
    static class TestModule {
        @Provides @Singleton @Named(DataModule.IO_THREAD)
        public Scheduler provideIoThreadScheduler() {
            return Schedulers.immediate();
        }

        @Provides @Singleton
        public LocalCache provideLocalCache() {
            return new InMemoryCache();
        }
    }
}

File: app/src/test/java/io/github/hidroh/materialistic/data/SessionManagerTest.kt
/*
 * Copyright (c) 2018 Ha Duy Trung
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.hidroh.materialistic.data

import io.github.hidroh.materialistic.test.InMemoryCache
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import rx.schedulers.Schedulers

@RunWith(JUnit4::class)
class SessionManagerTest {
  private lateinit var manager: SessionManager

  @Before
  fun setUp() {
    val cache = InMemoryCache()
    cache.setViewed("1")
    cache.setViewed("2")
    manager = SessionManager(Schedulers.immediate(), cache)
  }

  @Test
  fun testIsViewedNull() {
    assertThat(manager.isViewed(null).toBlocking().single()).isFalse()
  }

  @Test
  fun testIsViewedTrue() {
    assertThat(manager.isViewed("1").toBlocking().single()).isTrue()
  }

  @Test
  fun testIsViewedFalse() {
    assertThat(manager.isViewed("-1").toBlocking().single()).isFalse()
  }

  @Test
  fun testView() {
    assertThat(manager.isViewed("3").toBlocking().single()).isFalse()
    manager.view("3")
    assertThat(manager.isViewed("3").toBlocking().single()).isTrue()
  }
}
